src/MiniJuvix/Translation/MonoJuvixToMiniC.hs

```diff
@@ -277,66 +401,200 @@ goFunctionClause Mono.FunctionClause {..} = (clauseCondition, returnStmt)
-    Mono.PatternVariable v -> [(v ^. Mono.nameText, ctorField)]
-    Mono.PatternConstructorApp Mono.ConstructorApp {..} ->
-      goConstructorApp ctorField _constrAppConstructor _constrAppParameters
-    Mono.PatternWildcard {} -> []
+    returnStmt :: Sem r (Statement, [Function])
+    returnStmt = do
+      bindings <- buildPatternInfoTable argTyps clause
+      (decls :: [Function], clauseResult) <- runOutputList (runReader bindings (goExpression (clause ^. Mono.clauseBody)))
+      return (StatementReturn (Just clauseResult), decls)
+
+genClosure :: ClosureInfo -> [Function]
+genClosure ClosureInfo {..} =
+  [ Function
+      { _funcSig =
+          FunctionSig
+            { _funcReturnType = returnType ^. typeDeclType,
+              _funcIsPtr = returnType ^. typeIsPtr,
+              _funcQualifier = None,
+              _funcName = funName,
+              _funcArgs = namedArgs asFunArg (declFunctionPtrType : argTypes)
+            },
+        _funcBody =
+          [ returnStatement
+              ( functionCall
+                  (ExpressionVar _closureRootName)
+                  (ExpressionVar <$> take (length argTypes) (drop 1 funArgs))
+              )
+          ]
+      },
+    Function
+      { _funcSig =
+          FunctionSig
+            { _funcReturnType = declFunctionType,
+              _funcIsPtr = True,
+              _funcQualifier = None,
+              _funcName = asNew funName,
+              _funcArgs = []
+            },
+        _funcBody =
+          [ BodyDecl
+              ( Declaration
+                  { _declType = declFunctionType,
+                    _declIsPtr = True,
+                    _declName = Just localName,
+                    _declInitializer = Just $ ExprInitializer (mallocSizeOf Str.minijuvixFunctionT)
+                  }
+              ),
+            BodyStatement
+              ( StatementExpr
+                  ( ExpressionAssign
+                      ( Assign
+                          { _assignLeft = memberAccess Pointer (ExpressionVar localName) "fun",
+                            _assignRight =
+                              castToType
+                                ( CDeclType
+                                    { _typeDeclType = uIntPtrType,
+                                      _typeIsPtr = False
+                                    }
+                                )
+                                (ExpressionVar funName)
+                          }
+                      )
+                  )
+              ),
+            returnStatement (ExpressionVar localName)
+          ]
+      }
+  ]
   where
-    asConstructor :: Expression
-    asConstructor = functionCall (ExpressionVar (asCast (mkName n))) [arg]
+    returnType :: CDeclType
+    returnType = _closureFunType ^. cFunReturnType
+    argTypes :: [CDeclType]
+    argTypes = _closureFunType ^. cFunArgTypes
+    localName :: Text
+    localName = "f"
+    funName :: Text
+    funName = asFun _closureRootName
```

Comment:
For readability, consider moving this up with `let.in`. It feels to me a bit lost.